README.md
# everylog_node_client

HISTORY.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [1.1.0-beta.0](https://github.com/everylogsaas/everylog_node_client/compare/v1.0.1...1.1.0-beta.0) - 2023-04-28

### Commits

- updated HISTORY [`81bd506`](https://github.com/everylogsaas/everylog_node_client/commit/81bd5064acbca6576bb12447c4b4f6d8a1b6246e)
- updates v1.1.0 [`63f138c`](https://github.com/everylogsaas/everylog_node_client/commit/63f138c400df21515a96ebf7ad4cb312bca0d36e)
- add icon-channels-properties [`ff10559`](https://github.com/everylogsaas/everylog_node_client/commit/ff10559429f5da4e9f01509092ac48a64addd5b1)
- deleted CHANGELOG.md [`72511c9`](https://github.com/everylogsaas/everylog_node_client/commit/72511c945ea27d2f58b4241e8c68e9c488d8c748)

## [v1.0.1](https://github.com/everylogsaas/everylog_node_client/compare/v1.0.0...v1.0.1) - 2023-04-21

### Commits

- updates: [`26ed684`](https://github.com/everylogsaas/everylog_node_client/commit/26ed684244e039dce3ba0a06d249f9cbdf9bc92f)

## v1.0.0 - 2023-04-20

### Commits

- everylog client [`a849575`](https://github.com/everylogsaas/everylog_node_client/commit/a849575b180cad02fedf157038048ccb801a20b1)
- Initial commit [`bc04cd0`](https://github.com/everylogsaas/everylog_node_client/commit/bc04cd0f1f4578ff53f9817bf1458be2abbebef4)
- groups [`302aaa6`](https://github.com/everylogsaas/everylog_node_client/commit/302aaa66c0e6c4e90610447bcd3df69cb0b23cf1)
- updated package.json with details for publishing [`654351b`](https://github.com/everylogsaas/everylog_node_client/commit/654351bfe70d72b86732e54fc926c015e2ba7c8c)


src/index.js
import https from 'https';
import {
  validateNotificationOptions,
  checkApiKeyAndProjectId,
} from './utils';
import { EveryLogNotificationError } from './errors/notificationError';
import { EveryLogGenericError } from './errors/genericError';

export class EveryLogClient {
  /**
   *
   * @namespace settings
   * @property {Object} settings
   * @property {string} settings.apiKey
   * @property {string} settings.projectId - String, mandatory, the project unique identifier
   *
   * @namespace options
   * @property {string} protocol   - default: "https:"
   * @property {string} port       - default: "443"
   * @property {string} hostname   - default: "api.everylog.io"
   * @property {string} path       - default: "/api/v1/log-entries"
   * @property {string} method     - default: "POST"
   *
   */
  constructor(apiKey, projectId, {
    protocol = 'https:', port = 443, hostname = 'api.everylog.io', path = '/api/v1/log-entries', method = 'POST'
  } = {}) {
    checkApiKeyAndProjectId({ apiKey, projectId });
    this.apiKey = apiKey;
    this.projectId = projectId;
    this.protocol = protocol;
    this.port = port;
    this.hostname = hostname;
    this.path = path;
    this.method = method;
    this.headers = {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.apiKey}`,
    };
  }

  /**
  *
  * @namespace  notification
  * @property {object} notification
  * @property {string} notification.title - the title of the entry, length 50
  * @property {string} notification.summary - a quick summary, length 100
  * @property {object} notification.body -  the full body to log
  * @property {array[string]} [notification.tags] - useful to better filter entries
  * @property {string} [notification.links] - string as url, an external url to point
  * @property {boolean} [notification.push] - boolean, true to receive a push notification
  * @property {array[string]} [notification.groups] - useful to better send notifications
  * @property {array[string]} [notification.externalChannels] - send notification to externalChannel
  * @property {string} [notification.icon] - emoji icon fot notification
  * @property {object} [notification.properties] - useful to filter notifications by properties
  *
  *
  */
  async notify({
    title,
    summary,
    body,
    link = null,
    push = false,
    tags = [],
    groups = [],
    externalChannels = [],
    icon = null,
    properties = null
  }) {
    validateNotificationOptions({
      title, summary, body, link, push, tags, groups, externalChannels, icon, properties
    });

    const data = JSON.stringify({
      projectId: this.projectId,
      title,
      summary,
      body,
      ...(link !== null && { link }),
      ...(push !== null && { push }),
      ...(tags !== null && { tags }),
      ...(groups !== null && { groups })
    });

    const option = {
      protocol: this.protocol,
      port: this.port,
      hostname: this.hostname,
      path: this.path,
      method: this.method,
      headers: {
        ...this.headers,
        'Content-Length': Buffer.byteLength(data)
      }
    };

    return new Promise((resolve, reject) => {
      const req = https.request(option, (res) => {
        let resData = '';
        const { statusCode } = res;

        res.on('data', (chunk) => {
          resData += chunk;
        });

        res.on('end', () => {
          const parsedData = JSON.parse(resData);
          if (res.statusCode >= 200 && res.statusCode <= 299) {
            resolve({ response: parsedData, statusCode });
          } else {
            reject(new EveryLogNotificationError(parsedData.message, statusCode));
          }
        });
      })
        .on('error', (err) => {
          reject(new EveryLogGenericError(err.message));
        });
      req.write(data);
      req.end();
    });
  }
}

export default EveryLogClient;


src/utils.js
const isValidHttpUrl = (string) => {
  let url;
  try {
    url = new URL(string);
  } catch (_) {
    return false;
  }
  return url.protocol === 'http:' || url.protocol === 'https:';
};

export const validateNotificationOptions = (notificationOptions) => {
  if (!notificationOptions.title) {
    throw new Error('title required');
  } else if (notificationOptions.title.length > 50 || typeof notificationOptions.title !== 'string') {
    throw new Error('invalid title');
  }

  if (!notificationOptions.summary) {
    throw new Error('summary required');
  } else if (notificationOptions.summary.length > 100 || typeof notificationOptions.summary !== 'string') {
    throw new Error('invalid summary');
  }

  if (!notificationOptions.body) {
    throw new Error('body required');
  } else if (typeof notificationOptions.body !== 'string') {
    throw new Error('invalid body');
  }

  if (notificationOptions.push !== null && notificationOptions.push !== undefined) {
    if (typeof notificationOptions.push !== 'boolean') {
      throw new Error('invalid push');
    }
  }

  if (notificationOptions.link) {
    if (!isValidHttpUrl(notificationOptions.link)) {
      throw new Error('link is not a valid url');
    }
  }

  if (notificationOptions.tags) {
    if (!Array.isArray(notificationOptions.tags)) {
      throw new Error('tags is not an array');
    } else {
      const checkArrayType = notificationOptions.tags.find((element) => typeof element !== 'string');
      if (checkArrayType !== undefined) {
        throw new Error('invalid array element');
      }
    }
  }

  if (notificationOptions.groups) {
    if (!Array.isArray(notificationOptions.groups)) {
      throw new Error('groups is not an array');
    } else {
      const checkArrayType = notificationOptions.groups.find((element) => typeof element !== 'string');
      if (checkArrayType !== undefined) {
        throw new Error('invalid array groups');
      }
    }
  }

  if (notificationOptions.externalChannels) {
    if (!Array.isArray(notificationOptions.externalChannels)) {
      throw new Error('externalChannels is not an array');
    } else {
      const checkArrayType = notificationOptions.externalChannels.find((element) => typeof element !== 'string');
      if (checkArrayType !== undefined) {
        throw new Error('invalid array externalChannels');
      }
    }
  }

  if (notificationOptions.icon) {
    if (typeof notificationOptions.icon !== 'string') {
      throw new Error('invalid icon');
    }
  }

  if (notificationOptions.properties) {
    if (typeof notificationOptions.properties !== 'object') {
      throw new Error('invalid properties type');
    }
  }
};

export const checkApiKeyAndProjectId = (settings) => {
  if (!settings.apiKey) {
    throw new Error('Api Key required');
  } else if (settings.apiKey.length !== 36 || typeof settings.apiKey !== 'string') {
    throw new Error('invalid api key');
  }

  if (!settings.projectId) {
    throw new Error('projectId required');
  } else if (settings.projectId.length > 20 || typeof settings.projectId !== 'string') {
    throw new Error('invalid projectId');
  }
};


src/client.js
import https from "https"
import http from "http"
import { checkNotifications, checkOptions, checkSettings } from "./utils.js";

export default class EverylogClient {
  /**
   * 
   * @namespace settings
   * @property {Object} settings 
   * @property {string} settings.apiKey    - length: 36, mandatory, everylog account unique indentifier
   * @property {string} settings.projectId - length: 20, mandatory, the project unique identifier
   * 
   * @namespace options
   * @property {string} hostname   - default: "api.everylog.io"
   * @property {string} path       - default: "/api/v1/log-entries"
   * @property {string} method     - default: "POST"
   * @property {object} headers    - default: { 'Content-Type': 'application/json', 'Authorization': `Bearer {apiKey}`}
   * @property {number} port       - default: 443  
   * 
   */
  constructor(apiKey, projectId, { hostname = "api.everylog.io", path = "/api/v1/log-entries", method = "POST", port = 443 } = {}) {

    checkSettings({ apiKey, projectId })
    checkOptions({ hostname, path, method })
    this.apiKey = apiKey;
    this.projectId = projectId;
    this.protocol = "https:";
    this.hostname = hostname;
    this.path = path;
    this.method = method;
    this.port = port
    this.headers = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${this.apiKey}`,
    }
  }

  /**
  * 
  * @namespace  notification
  * @property {object} notification         
  * @property {string} notification.title - length: 50, mandatory, the title of the entry
  * @property {string} notification.summary - length: 100, mandatory, a quick summary
  * @property {object} notification.body - string, mandatory, the full body to log
  * @property {array[string]} notification.tags - array of strings, optional, useful to better filter entries
  * @property {string} notification.links - string as url, optional, an external url to point
  * @property {boolean} notification.push - boolean, optional, true to receive a push notification
  * @property {array[string]} notification.groups - array of strngs, optional, useful to better send notifications
  * @property {array[string]} notification.externalChannels - array of strings, optional, send notification to externalChannel
  * @property {string} notification.icon -string as icon, optional, icon fot notification
  * @property {object} notification.properties -object {key: "value"}, optional, useful to filter notifications by properties
  * 
  */
  async notify({ title, summary, body, link = null, push = false, tags = [], groups = [], externalChannels = [], icon = null, properties = null }) {

    checkNotifications({ title, summary, body, link, push, tags, groups, externalChannels, icon, properties })

    const data = JSON.stringify({
      projectId: this.projectId,
      title: title,
      summary: summary,
      body: body,
      ...(link != null && { link: link }),
      ...(push != null && { push: push }),
      ...(tags != null && { tags: tags }),
      ...(groups != null && { groups: groups }),
      ...(externalChannels != null && { externalChannels: externalChannels }),
      ...(icon != null && { icon: icon }),
      ...(properties != null && { properties: properties }),
    })

    console.log(data);

    let option = {
      //protocol: this.protocol,
      hostname: this.hostname,
      path: this.path,
      method: this.method,
      headers: {
        // 'Authorization': this.headers.Authorization + this.apiKey,
        // 'Content-Type': this.headers["Content-Type"],
        ...this.headers,
        // 'Content-Length': data.length,
      },
    }

    if (this.port === 443) {
      option.protocol = "https:"

    } else {
      option.port = this.port
    }


    return new Promise((resolve, reject) => {
      const req = (this.port === 443 ? https : http).request(option, res => {
        let resData = ''
        let statusCode = res.statusCode

        res.on("data", chunk => {
          resData += chunk
        });

        res.on("end", () => {
          const parsedData = JSON.parse(resData)
          if (res.statusCode >= 200 && res.statusCode <= 299 || res.statusCode == 401) {
            resolve({ response: parsedData, statusCode })
          } else {
            reject({ error: parsedData.message, statusCode })
          }
        })
      })
        .on("error", err => {
          reject({ error: err })
        })
      req.write(data)
      req.end()
    })
  }
};
